### PR TITLE
add sdf crypto interfaces

### PR DESCRIPTION
**Pull request by MaggieNgWu**

### CODE REVIEW

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+namespace crypto
+{
+/**
+ *  SDFCryptoProvider suply SDF function calls
+ *  Singleton
+ */
+class SDFCryptoProvider
+{
+private:
+    static SDFCryptoProvider _instance;
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFCryptoProvider.h:35]: Remove this.

Comment (bxq2011hust) [inline libdevcrypto/sdf/SDFCryptoProvider.h:35]: This will compile failed

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     * Generate key
+     * Return error code
+     */
+    unsigned int KeyGen(std::string algorithm, Key* key);
+
+    /**
+     * Sign
+     */
+    unsigned int Sign(Key const* privateKey, std::string algorithm, unsigned char const* digest,
+        int digestLen, unsigned char* signature, int* signatureLen);
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFCryptoProvider.h:69]: `unsigned char const*` -> `std::shared_ptr<bytes>` or `bytesConstRef`

Comment (cyjseagull) [reply libdevcrypto/sdf/SDFCryptoProvider.h:69]: `int* signatureLen`->`size_t signatureLen`

File: libdevcrypto/sdf/SDFSM2Signature.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "Signature.h"
+#include "libdevcore/RLP.h"
+#include <vector>
+
+namespace dev
+{
+namespace crypto
+{
+struct SDFSM2Signature : public Signature
```
Comment (bxq2011hust) [inline libdevcrypto/sdf/SDFSM2Signature.h:32]: just use SM2Signature

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Make hash
+     */
+    unsigned int Hash(Key const* key, char const* message, std::string algorithm,
+        int const messageLen, char* digest, int* digestLen);
+
+    /**
+     * Encrypt
+     */
+    unsigned int Encrypt(Key const* publicKey, std::string algorithm, unsigned char const* plantext,
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFCryptoProvider.h:86]: `Encrypt` to `asymmetricEncrypt`

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        int const plantextLen, unsigned char* cyphertext, int* cyphertextLen);
+
+    /**
+     * Decrypt
+     */
+    unsigned int Decrypt(Key const* privateKey, std::string algorithm,
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFCryptoProvider.h:92]: `Decrypt` to `asymmetricDecrypt`

File: libdevcrypto/sdf/SDFSM2Signature.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+};
+}  // namespace crypto
+
+class KeyPair;
+std::shared_ptr<crypto::Signature> SDFSM2Sign(KeyPair const& _keyPair, const h256& _hash);
+bool SDFSM2Verify(h512 const& _pubKey, std::shared_ptr<crypto::Signature> _sig, const h256& _hash);
+h512 SDFSM2Recover(std::shared_ptr<crypto::Signature> _sig, const h256& _hash);
+std::shared_ptr<crypto::Signature> SDFSM2SignatureFromRLP(RLP const& _rlp, size_t _start);
+std::shared_ptr<crypto::Signature> SDFSM2SignatureFromBytes(std::vector<unsigned char> _data);
+std::pair<bool, std::vector<unsigned char> > Recover(bytesConstRef _in);
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFSM2Signature.h:43]: `Recover`->`SDFSM2Recover`

File: libdevcrypto/sdf/SDFSM3Hash.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * @date 2021-02-01
+ */
+
+#pragma once
+#include <libdevcore/FixedHash.h>
+#include <libdevcore/vector_ref.h>
+#include <string>
+namespace dev
+{
+inline SecureFixedHash<32> SDFSM3Secure(std::string const& _input);
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFSM3Hash.h:28]: `SecureFixedHash<32>`->`h256`

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    /**
+     * Sign
+     */
+    unsigned int Sign(Key const* privateKey, std::string algorithm, unsigned char const* digest,
+        int digestLen, unsigned char* signature, int* signatureLen);
+
+    /**
+     * Verify signature
+     */
+    unsigned int Verify(Key const* publicKey, std::string algorithm, unsigned char const* digest,
```
Comment (bxq2011hust) [inline libdevcrypto/sdf/SDFCryptoProvider.h:74]: the algorithm can be in Key?

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        int digestLen, unsigned char const* signature, int* signatureLen, bool result);
```
Comment (bxq2011hust) [inline libdevcrypto/sdf/SDFCryptoProvider.h:75]: what's the difference between result and return value?

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+     * Detect all crypto devices
+     * Open each device, and start a session
+     */
+    SDFCryptoProvider(/* args */);
+    /**
+     * Close sessions and close devices.
+     */
+    ~SDFCryptoProvider();
+
+    class Key;
```
Comment (cyjseagull) [inline libdevcrypto/sdf/SDFCryptoProvider.h:53]: Where is the definition of the `Key`

File: libdevcrypto/sdf/SDFCryptoProvider.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+ */
+class SDFCryptoProvider
+{
+private:
+    static SDFCryptoProvider _instance;
+
+    /**
+     * Get session from session pool
+     */
+    static void getSession(void* handle, unsigned int* code);
```
Comment (bxq2011hust) [inline libdevcrypto/sdf/SDFCryptoProvider.h:40]: why we need a pool of sessions?

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 9 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h]: Result is about whether the signature is valid or not.
Return value is an error code returned by hardware secure module.

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 4 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h]: Get

File: libdevcrypto/sdf/SDFSM3Hash.h (hunk 7 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFSM3Hash.h:28]: SecureFixedHash<32> is used here to keep consistent with SM3Hash.h.

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 10 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h:53]: Soon supply in the next pr.

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 11 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h:40]: The host may connect more than one HSM device, and each device can create multiple session. So we need a session pool to manage them.

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 1 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h:35]: Will remove this.

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 2 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h]: I prefer use ``unsigned char const*`` and ``unsigned int``, which are consistent with SDF interface.

File: libdevcrypto/sdf/SDFSM2Signature.h (hunk 6 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFSM2Signature.h:43]: Get, thanks!

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 1 above)
Comment (MaggieNgWu) [reply libdevcrypto/sdf/SDFCryptoProvider.h:35]: Get

File: libdevcrypto/sdf/SDFCryptoProvider.h (hunk 11 above)
Comment (bxq2011hust) [reply libdevcrypto/sdf/SDFCryptoProvider.h:40]: I think we can keep the session inside SDFCryptoProvider, the code call SDFCryptoProvider doesn't need to care hardware.
besides the parameter of  getSession can't get info out, should use pointer of pointer

**Review by bxq2011hust — APPROVED** [review]

**Review by cyjseagull — APPROVED** [review]

**Pull request merged**